File: app/api/notifications.py
"""
FastAPI routes for the notification backend: inbox, push subscriptions,
preferences, and (owner/co-owner only) daily-report scheduling.
"""
from __future__ import annotations

import json

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies.auth import AuthContext, get_auth_context, require_roles
from app.schemas.notification import (
    DailyReportScheduleRead,
    DailyReportScheduleUpdate,
    NotificationListResponse,
    NotificationMarkReadRequest,
    NotificationPreferenceRead,
    NotificationPreferenceUpdate,
    NotificationRead,
    PushSubscriptionRegisterRequest,
    PushSubscriptionUnregisterRequest,
)
from app.services import notification_service
from app.services.permission_service import ROLE_CO_OWNER, ROLE_OWNER

router = APIRouter(prefix="/api/notifications", tags=["Notifications"])


@router.get("", response_model=NotificationListResponse)
def list_my_notifications(
    unread_only: bool = Query(default=False),
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
    db: Session = Depends(get_db),
    context: AuthContext = Depends(get_auth_context),
):
    items = notification_service.list_notifications(
        db, hotel_id=context.hotel_id, user_id=context.user_id,
        unread_only=unread_only, limit=limit, offset=offset,
    )
    unread_count = notification_service.unread_notification_count(db, hotel_id=context.hotel_id, user_id=context.user_id)
    return NotificationListResponse(items=items, unread_count=unread_count)


@router.post("/{notification_id}/read", response_model=NotificationRead)
def mark_read(
    notification_id: int,
    data: NotificationMarkReadRequest = NotificationMarkReadRequest(),
    db: Session = Depends(get_db),
    context: AuthContext = Depends(get_auth_context),
):
    try:
        notification = notification_service.mark_notification_read(
            db, hotel_id=context.hotel_id, user_id=context.user_id,
            notification_id=notification_id, read=data.read,
        )
    except notification_service.NotificationServiceError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    db.commit()
    db.refresh(notification)
    return notification


@router.post("/read-all")
def mark_all_read(
    db: Session = Depends(get_db),
    context: AuthContext = Depends(get_auth_context),
):
    count = notification_service.mark_all_notifications_read(db, hotel_id=context.hotel_id, user_id=context.user_id)
    db.commit()
    return {"marked_read": count}


@router.post("/push-subscriptions", status_code=status.HTTP_201_CREATED)
def register_push_subscription(
    data: PushSubscriptionRegisterRequest,
    db: Session = Depends(get_db),
    context: AuthContext = Depends(get_auth_context),
):
    notification_service.register_push_subscription(
        db, hotel_id=context.hotel_id, user_id=context.user_id,
        endpoint=data.endpoint, p256dh=data.p256dh, auth=data.auth,
    )
    db.commit()
    return {"registered": True}


@router.post("/push-subscriptions/unregister")
def unregister_push_subscription(
    data: PushSubscriptionUnregisterRequest,
    db: Session = Depends(get_db),
    context: AuthContext = Depends(get_auth_context),
):
    removed = notification_service.unregister_push_subscription(
        db, hotel_id=context.hotel_id, user_id=context.user_id, endpoint=data.endpoint,
    )
    db.commit()
    return {"removed": removed}


@router.get("/preferences", response_model=list[NotificationPreferenceRead])
def list_preferences(
    db: Session = Depends(get_db),
    context: AuthContext = Depends(get_auth_context),
):
    return notification_service.list_notification_preferences(db, hotel_id=context.hotel_id, user_id=context.user_id)


@router.put("/preferences", response_model=NotificationPreferenceRead)
def update_preference(
    data: NotificationPreferenceUpdate,
    db: Session = Depends(get_db),
    context: AuthContext = Depends(get_auth_context),
):
    try:
        row = notification_service.set_notification_preference(
            db,
            hotel_id=context.hotel_id,
            user_id=context.user_id,
            event_type=data.event_type,
            in_app_enabled=data.in_app_enabled,
            push_enabled=data.push_enabled,
            email_enabled=data.email_enabled,
            quiet_hours_start=data.quiet_hours_start,
            quiet_hours_end=data.quiet_hours_end,
            clear_quiet_hours=data.clear_quiet_hours,
        )
    except notification_service.NotificationServiceError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc))
    db.commit()
    db.refresh(row)
    return row


def _schedule_to_read(hotel_id: int, row) -> DailyReportScheduleRead:
    if row is None:
        return DailyReportScheduleRead(
            hotel_id=hotel_id, hour=7, recipient_user_ids=[], recipient_roles=[],
            channels=["in_app"], sections=notification_service.DEFAULT_DAILY_REPORT_SECTIONS, enabled=False,
        )
    return DailyReportScheduleRead(
        hotel_id=hotel_id,
        hour=row.hour,
        recipient_user_ids=json.loads(row.recipient_user_ids or "[]"),
        recipient_roles=json.loads(row.recipient_roles or "[]"),
        channels=json.loads(row.channels or '["in_app"]'),
        sections=json.loads(row.sections or "[]"),
        enabled=row.enabled,
    )


@router.get("/daily-report-schedule", response_model=DailyReportScheduleRead)
def get_daily_report_schedule(
    db: Session = Depends(get_db),
    context: AuthContext = Depends(require_roles(ROLE_OWNER, ROLE_CO_OWNER)),
):
    row = notification_service.get_daily_report_schedule(db, hotel_id=context.hotel_id)
    return _schedule_to_read(context.hotel_id, row)


@router.put("/daily-report-schedule", response_model=DailyReportScheduleRead)
def update_daily_report_schedule(
    data: DailyReportScheduleUpdate,
    db: Session = Depends(get_db),
    context: AuthContext = Depends(require_roles(ROLE_OWNER, ROLE_CO_OWNER)),
):
    try:
        row = notification_service.upsert_daily_report_schedule(
            db,
            hotel_id=context.hotel_id,
            hour=data.hour,
            recipient_user_ids=data.recipient_user_ids,
            recipient_roles=data.recipient_roles,
            channels=data.channels,
            sections=data.sections,
            enabled=data.enabled,
        )
    except notification_service.NotificationServiceError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc))
    db.commit()
    return _schedule_to_read(context.hotel_id, row)

File: frontend/src/api/notifications.ts
import { apiFetch, type SessionLike } from "./client";

// Mirrors app/models/notification.py NotificationSeverityEnum.
export type NotificationSeverity = "info" | "warning" | "critical";

// Mirrors app/schemas/notification.py NotificationRead. Deliberately no
// `payload` field -- the backend never puts guest/reservation detail there
// (see app/services/notification_service.py's SAFE_PAYLOAD_KEYS discipline),
// and NotificationRead itself doesn't expose it, so the frontend must not
// assume any field beyond these exists. `entity_type`/`entity_id` are the
// only deep-link material available; whatever the target page needs beyond
// that id is re-fetched (and re-authorized) by that page itself.
export type NotificationItem = {
  id: number;
  hotel_id: number;
  recipient_user_id: number;
  event_type: string;
  severity: NotificationSeverity;
  title: string;
  body?: string | null;
  entity_type?: string | null;
  entity_id?: number | null;
  is_read: boolean;
  read_at?: string | null;
  created_at: string;
  expires_at?: string | null;
};

export type NotificationListResponse = {
  items: NotificationItem[];
  unread_count: number;
};

export const listNotifications = (
  params: { unreadOnly?: boolean; limit?: number; offset?: number } = {},
  session?: SessionLike
) => {
  const query = new URLSearchParams();
  if (params.unreadOnly) query.set("unread_only", "true");
  query.set("limit", String(params.limit ?? 50));
  query.set("offset", String(params.offset ?? 0));
  return apiFetch<NotificationListResponse>(`/api/notifications?${query.toString()}`, { session });
};

export const markNotificationRead = (notificationId: number, read = true, session?: SessionLike) =>
  apiFetch<NotificationItem>(`/api/notifications/${notificationId}/read`, {
    method: "POST",
    data: { read },
    session
  });

export const markAllNotificationsRead = (session?: SessionLike) =>
  apiFetch<{ marked_read: number }>("/api/notifications/read-all", { method: "POST", session });

export type PushSubscriptionPayload = { endpoint: string; p256dh: string; auth: string };

export const registerPushSubscription = (payload: PushSubscriptionPayload, session?: SessionLike) =>
  apiFetch<{ registered: boolean }>("/api/notifications/push-subscriptions", {
    method: "POST",
    data: payload,
    session
  });

export const unregisterPushSubscription = (endpoint: string, session?: SessionLike) =>
  apiFetch<{ removed: boolean }>("/api/notifications/push-subscriptions/unregister", {
    method: "POST",
    data: { endpoint },
    session
  });

// Mirrors app/schemas/notification.py NotificationPreferenceRead.
// `event_type === ""` is the hotel-wide default row (see the backend model's
// own docstring) -- every other event_type row overrides it individually.
export type NotificationPreference = {
  event_type: string;
  in_app_enabled: boolean;
  push_enabled: boolean;
  email_enabled: boolean;
  quiet_hours_start?: number | null;
  quiet_hours_end?: number | null;
};

export type NotificationPreferenceUpdate = {
  event_type?: string;
  in_app_enabled?: boolean;
  push_enabled?: boolean;
  email_enabled?: boolean;
  quiet_hours_start?: number | null;
  quiet_hours_end?: number | null;
  clear_quiet_hours?: boolean;
};

export const listNotificationPreferences = (session?: SessionLike) =>
  apiFetch<NotificationPreference[]>("/api/notifications/preferences", { session });

export const updateNotificationPreference = (payload: NotificationPreferenceUpdate, session?: SessionLike) =>
  apiFetch<NotificationPreference>("/api/notifications/preferences", { method: "PUT", data: payload, session });

// Mirrors app/schemas/notification.py DailyReportScheduleRead/Update. Owner
// / co-owner only on the backend (require_roles(ROLE_OWNER, ROLE_CO_OWNER)).
export type DailyReportSchedule = {
  hotel_id: number;
  hour: number;
  recipient_user_ids: number[];
  recipient_roles: string[];
  channels: string[];
  sections: string[];
  enabled: boolean;
};

export type DailyReportScheduleUpdate = {
  hour: number;
  recipient_user_ids: number[];
  recipient_roles: string[];
  channels: string[];
  sections?: string[] | null;
  enabled: boolean;
};

export const getDailyReportSchedule = (session?: SessionLike) =>
  apiFetch<DailyReportSchedule>("/api/notifications/daily-report-schedule", { session });

export const updateDailyReportSchedule = (payload: DailyReportScheduleUpdate, session?: SessionLike) =>
  apiFetch<DailyReportSchedule>("/api/notifications/daily-report-schedule", {
    method: "PUT",
    data: payload,
    session
  });
